### PR TITLE
Use NSDecimalString(_:_) for Decimal -> String conversion.

### PR DESCRIPTION
**Pull request by spevans**
- This allows using an optional Locale to localise the output.

- Update Decimal.description, NSDecimalNumber.description and
  NSDecimalNumber.description(withLocale) to call NSDecimalString().

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci test linux

File: Foundation/Decimal.swift [hunk 1]
```diff
@@ -1691,10 +1639,71 @@ public func NSDecimalMultiplyByPowerOf10(_ result: UnsafeMutablePointer<Decimal>
...
+    if copy._isNegative != 0 {
+        i -= 1
+        buffer[i] = MINUS
+    }
+    let result = String(cString: Array(buffer.suffix(from: i)))
+
+    var decimalSeparator: String?
+    if let locale = locale as? Locale {
+        decimalSeparator = locale.decimalSeparator
+    } else if let dictionary = locale as? Dictionary<NSLocale.Key, String> {
```
Comment (millenomi) [inline Foundation/Decimal.swift]: Should this be `[NSLocale.Key: Any]`?

Note that unlike Darwin, `NSLocale.Key` will not automatically bridge from `String` or `NSString`. Should we have a branch for that too?

File: TestFoundation/TestDecimal.swift [hunk 2]
```diff
@@ -9,35 +9,6 @@
 
 class TestDecimal: XCTestCase {
 
-    static var allTests : [(String, (TestDecimal) -> () throws -> Void)] {
```
Comment (millenomi) [inline TestFoundation/TestDecimal.swift]: Thanks for moving this.

**Comment by spevans** [comment]
@swift-ci test linux

**Comment by spevans** [comment]
@swift-ci test linux

**Comment by spevans** [comment]
@swift-ci test linux

File: Foundation/Decimal.swift (hunk 1 above)
Comment (spevans) [reply Foundation/Decimal.swift]: I added a branch for `[String: String]`. 

For the `[NSLocale.Key: Any]` case wouldn't that then mean `as!` casting the value to `String`? It that what you want it changed to?

Comment (millenomi) [reply Foundation/Decimal.swift]: No, the current patch is good as-is.

Comment (millenomi) [reply Foundation/Decimal.swift]: (After adding the branch.)

**Pull request merged**